File: apps/desktop/src-tauri/src/db/mysql.rs
use crate::error::AppResult;
use crate::model::{
    ColumnInfo, ConnectionOverview, DatabaseSummary, ForeignKeyInfo, IndexInfo, PageOptions,
    QueryResult, QueryResultColumn, RelationInfo, RelationKind, ResolvedConnection, SchemaInfo,
    TableDetails, TestConnectionResult, QUERY_LIMIT,
};
use mysql_async::prelude::*;
use mysql_async::{Conn, Opts, OptsBuilder, Row, Value as MyValue};
use serde_json::{json, Value};
use std::time::Instant;

async fn connect(c: &ResolvedConnection) -> AppResult<Conn> {
    let opts: Opts = OptsBuilder::default()
        .ip_or_hostname(c.host.clone())
        .tcp_port(c.port)
        .user(Some(c.username.clone()))
        .pass(Some(c.password.clone()))
        .db_name(Some(c.database.clone()))
        .into();
    Ok(Conn::new(opts).await?)
}

fn ident(name: &str) -> String {
    format!("`{}`", name.replace('`', "``"))
}

pub async fn test(c: &ResolvedConnection) -> AppResult<TestConnectionResult> {
    let start = Instant::now();
    match connect(c).await {
        Ok(mut conn) => {
            let res: Result<Option<i64>, _> = conn.query_first("SELECT 1").await;
            let ok = res.is_ok();
            let latency = start.elapsed().as_millis() as u64;
            let _ = conn.disconnect().await;
            if ok {
                Ok(TestConnectionResult {
                    ok: true,
                    message: "Conexión exitosa".into(),
                    latency_ms: Some(latency),
                })
            } else {
                Ok(TestConnectionResult {
                    ok: false,
                    message: res.err().map(|e| e.to_string()).unwrap_or_default(),
                    latency_ms: None,
                })
            }
        }
        Err(e) => Ok(TestConnectionResult {
            ok: false,
            message: e.to_string(),
            latency_ms: None,
        }),
    }
}

pub async fn list_schemas(c: &ResolvedConnection) -> AppResult<Vec<SchemaInfo>> {
    let mut conn = connect(c).await?;
    let rows: Vec<String> = conn.query("SHOW DATABASES").await?;
    let _ = conn.disconnect().await;
    let sys: std::collections::HashSet<&str> =
        ["information_schema", "mysql", "performance_schema", "sys"].into_iter().collect();
    Ok(rows
        .into_iter()
        .map(|name| {
            let is_system = sys.contains(name.as_str());
            SchemaInfo { name, is_system }
        })
        .collect())
}

pub async fn list_relations(
    c: &ResolvedConnection,
    schema: Option<&str>,
) -> AppResult<Vec<RelationInfo>> {
    let mut conn = connect(c).await?;
    let target = schema.map(String::from).unwrap_or_else(|| c.database.clone());
    let rows: Vec<(String, String, String)> = conn
        .exec(
            "SELECT table_schema, table_name, table_type FROM information_schema.tables
             WHERE table_schema = ? ORDER BY table_name",
            (target,),
        )
        .await?;
    let _ = conn.disconnect().await;
    Ok(rows
        .into_iter()
        .map(|(s, n, t)| {
            let kind = if t == "VIEW" {
                RelationKind::View
            } else {
                RelationKind::Table
            };
            RelationInfo {
                schema: s,
                name: n,
                kind,
            }
        })
        .collect())
}

pub async fn get_connection_overview(c: &ResolvedConnection) -> AppResult<ConnectionOverview> {
    let mut conn = connect(c).await?;
    let sys: std::collections::HashSet<&str> =
        ["information_schema", "mysql", "performance_schema", "sys"]
            .into_iter()
            .collect();

    let meta: Option<(Option<String>, Option<String>, Option<String>, Option<String>, Option<u64>)> =
        conn.query_first(
            "SELECT VERSION(), DATABASE(), CURRENT_USER(), DATE_FORMAT(NOW(), '%Y-%m-%dT%H:%i:%s'), @@max_connections",
        )
        .await?;
    let (version, current_db, current_user, server_time, max_conn) =
        meta.unwrap_or((None, None, None, None, None));

    // Uptime via SHOW GLOBAL STATUS — wrap to survive missing privilege.
    let uptime: Option<u64> = match conn
        .query_first::<(String, String), _>("SHOW GLOBAL STATUS LIKE 'Uptime'")
        .await
    {
        Ok(Some((_, v))) => v.parse::<u64>().ok(),
        _ => None,
    };

    let db_rows: Vec<(String, Option<u64>, u64, Option<String>)> = conn
        .query(
            "SELECT s.schema_name,
                    COALESCE(SUM(t.data_length + t.index_length), 0),
                    COUNT(t.table_name),
                    s.default_character_set_name
             FROM information_schema.schemata s
             LEFT JOIN information_schema.tables t ON t.table_schema = s.schema_name
             GROUP BY s.schema_name, s.default_character_set_name
             ORDER BY s.schema_name",
        )
        .await
        .unwrap_or_default();

    let active: Option<u64> = conn
        .query_first::<u64, _>(
            "SELECT COUNT(*) FROM information_schema.processlist",
        )
        .await
        .ok()
        .flatten();

    let _ = conn.disconnect().await;

    let mut databases = Vec::with_capacity(db_rows.len());
    let mut total: u64 = 0;
    for (name, size_bytes, rel_count, charset) in db_rows {
        let size = size_bytes.and_then(|s| if s > 0 { Some(s) } else { None });
        if let Some(s) = size {
            total = total.saturating_add(s);
        }
        let is_system = sys.contains(name.as_str());
        databases.push(DatabaseSummary {
            name,
            is_system,
            size_bytes: size,
            relation_count: Some(rel_count),
            details: charset,
        });
    }

    Ok(ConnectionOverview {
        driver: "mysql".to_string(),
        server_version: version.map(|v| format!("MySQL {}", v)),
        current_database: current_db,
        current_user,
        server_time,
        uptime_seconds: uptime,
        total_size_bytes: if total > 0 { Some(total) } else { None },
        active_connections: active,
        max_connections: max_conn,
        databases,
    })
}

pub async fn describe_table(
    c: &ResolvedConnection,
    schema: &str,
    name: &str,
) -> AppResult<TableDetails> {
    let mut conn = connect(c).await?;
    let rows: Vec<(String, String, String, String, Option<String>)> = conn
        .exec(
            "SELECT column_name, column_type, is_nullable, column_key, column_default
             FROM information_schema.columns
             WHERE table_schema = ? AND table_name = ? ORDER BY ordinal_position",
            (schema, name),
        )
        .await?;
    // Indexes — group rows by index name.
    let idx_rows: Vec<(String, i64, String, u32)> = conn
        .exec(
            "SELECT index_name, non_unique, column_name, seq_in_index
             FROM information_schema.statistics
             WHERE table_schema = ? AND table_name = ?
             ORDER BY index_name, seq_in_index",
            (schema, name),
        )
        .await?;
    let mut idx_map: std::collections::BTreeMap<String, IndexInfo> =
        std::collections::BTreeMap::new();
    for (idx_name, non_unique, col_name, _seq) in idx_rows {
        let entry = idx_map
            .entry(idx_name.clone())
            .or_insert_with(|| IndexInfo {
                name: idx_name.clone(),
                columns: Vec::new(),
                unique: non_unique == 0,
                primary: idx_name == "PRIMARY",
            });
        entry.columns.push(col_name);
    }
    let mut indexes: Vec<IndexInfo> = idx_map.into_values().collect();
    indexes.sort_by(|a, b| {
        b.primary
            .cmp(&a.primary)
            .then(b.unique.cmp(&a.unique))
            .then(a.name.cmp(&b.name))
    });
    // Foreign keys — aggregate columns per constraint with GROUP_CONCAT.
    let fk_rows: Vec<(String, String, String, String, String, String, String)> = conn
        .exec(
            "SELECT rc.constraint_name,
                    GROUP_CONCAT(kcu.column_name ORDER BY kcu.ordinal_position),
                    rc.unique_constraint_schema,
                    rc.referenced_table_name,
                    GROUP_CONCAT(kcu.referenced_column_name ORDER BY kcu.ordinal_position),
                    rc.update_rule,
                    rc.delete_rule
             FROM information_schema.referential_constraints rc
             JOIN information_schema.key_column_usage kcu
               ON kcu.constraint_schema = rc.constraint_schema
              AND kcu.constraint_name = rc.constraint_name
             WHERE kcu.table_schema = ? AND kcu.table_name = ?
             GROUP BY rc.constraint_name, rc.unique_constraint_schema,
                      rc.referenced_table_name, rc.update_rule, rc.delete_rule
             ORDER BY rc.constraint_name",
            (schema, name),
        )
        .await?;
    let foreign_keys: Vec<ForeignKeyInfo> = fk_rows
        .into_iter()
        .map(|(fkname, local, rs, rt, refc, up, del)| ForeignKeyInfo {
            name: fkname,
            columns: local.split(',').map(|s| s.to_string()).collect(),
            referenced_schema: rs,
            referenced_table: rt,
            referenced_columns: refc.split(',').map(|s| s.to_string()).collect(),
            on_update: if up.is_empty() { None } else { Some(up) },
            on_delete: if del.is_empty() { None } else { Some(del) },
        })
        .collect();
    let _ = conn.disconnect().await;
    let columns = rows
        .into_iter()
        .map(|(col, ty, null, key, def)| ColumnInfo {
            name: col,
            data_type: ty,
            nullable: null == "YES",
            is_primary_key: key == "PRI",
            is_unique: key == "UNI",
            default: def,
        })
        .collect();
    Ok(TableDetails {
        schema: schema.into(),
        name: name.into(),
        kind: RelationKind::Table,
        columns,
        indexes,
        foreign_keys,
    })
}

pub async fn run_query(c: &ResolvedConnection, sql: &str) -> AppResult<QueryResult> {
    let start = Instant::now();
    let mut conn = connect(c).await?;
    let mut result = conn.query_iter(sql).await?;
    let mut all_columns: Vec<QueryResultColumn> = vec![];
    let mut rows_out: Vec<Vec<Value>> = vec![];
    let mut affected: Option<u64> = None;
    let mut truncated = false;

    while !result.is_empty() {
        let cols_now: Vec<QueryResultColumn> = result
            .columns_ref()
            .iter()
            .map(|c| QueryResultColumn {
                name: c.name_str().to_string(),
                data_type: format!("{:?}", c.column_type()),
            })
            .collect();
        if all_columns.is_empty() && !cols_now.is_empty() {
            all_columns = cols_now;
        }
        let collected: Vec<Row> = result.collect().await?;
        if collected.is_empty() {
            affected = Some(result.affected_rows());
        }
        for row in collected {
            if rows_out.len() >= QUERY_LIMIT {
                truncated = true;
                break;
            }
            let json_row: Vec<Value> = (0..row.len()).map(|i| myvalue_to_json(&row[i])).collect();
            rows_out.push(json_row);
        }
        if truncated {
            break;
        }
    }
    drop(result);
    let _ = conn.disconnect().await;

    Ok(QueryResult {
        row_count: rows_out.len(),
        columns: all_columns,
        rows: rows_out,
        affected_rows: affected,
        duration_ms: start.elapsed().as_millis() as u64,
        truncated,
    })
}

pub async fn fetch_table_data(
    c: &ResolvedConnection,
    schema: &str,
    name: &str,
    options: &PageOptions,
) -> AppResult<QueryResult> {
    let limit = options.limit.unwrap_or(100).min(QUERY_LIMIT as u32);
    let offset = options.offset.unwrap_or(0);
    let order = options
        .order_by
        .as_ref()
        .map(|cols| {
            let parts: Vec<String> = cols
                .iter()
                .map(|o| {
                    let dir = if o.direction.eq_ignore_ascii_case("desc") {
                        "DESC"
                    } else {
                        "ASC"
                    };
                    format!("{} {}", ident(&o.column), dir)
                })
                .collect();
            format!("ORDER BY {}", parts.join(", "))
        })
        .unwrap_or_default();
    let where_clause = options
        .where_clause
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(|s| format!("WHERE {}", s))
        .unwrap_or_default();
    let sql = format!(
        "SELECT * FROM {}.{} {} {} LIMIT {} OFFSET {}",
        ident(schema),
        ident(name),
        where_clause,
        order,
        limit,
        offset
    );
    run_query(c, &sql).await
}

fn myvalue_to_json(v: &MyValue) -> Value {
    match v {
        MyValue::NULL => Value::Null,
        MyValue::Bytes(b) => match std::str::from_utf8(b) {
            Ok(s) => Value::String(s.to_string()),
            Err(_) => json!({ "__binary": true, "bytes": b.len() }),
        },
        MyValue::Int(i) => json!(i),
        MyValue::UInt(u) => json!(u),
        MyValue::Float(f) => json!(f),
        MyValue::Double(f) => json!(f),
        MyValue::Date(y, m, d, h, mi, s, _) => {
            Value::String(format!("{:04}-{:02}-{:02} {:02}:{:02}:{:02}", y, m, d, h, mi, s))
        }
        MyValue::Time(_, _, h, m, s, _) => Value::String(format!("{:02}:{:02}:{:02}", h, m, s)),
    }
}

File: apps/web/src/server/db/mysql.ts
import mysql from "mysql2/promise";
import type { ColumnInfo, QueryResult, QueryResultColumn } from "@data-view/core";
import { quoteIdent as qiCore } from "@data-view/core";
import type {
  DriverAdapter,
  IterateTableOptions,
  ResolvedConnection,
  TableBatch,
} from "./types";

const QUERY_LIMIT = 5000;

async function withConn<T>(c: ResolvedConnection, fn: (conn: mysql.Connection) => Promise<T>): Promise<T> {
  const conn = await mysql.createConnection({
    host: c.host,
    port: c.port,
    user: c.username,
    password: c.password,
    database: c.database,
    ssl: c.ssl ? {} : undefined,
    connectTimeout: 10_000,
    rowsAsArray: true,
    dateStrings: true,
    multipleStatements: false,
  });
  try {
    return await fn(conn);
  } finally {
    await conn.end().catch(() => undefined);
  }
}

function ident(name: string) {
  return "`" + name.replace(/`/g, "``") + "`";
}

function normalizeCell(v: unknown) {
  if (v === null || v === undefined) return null;
  if (Buffer.isBuffer(v)) return { __binary: true as const, bytes: v.length };
  if (v instanceof Date) return v.toISOString();
  if (typeof v === "object") return JSON.stringify(v);
  return v as string | number | boolean;
}

export const mysqlDriver: DriverAdapter = {
  async testConnection(c) {
    const start = Date.now();
    try {
      await withConn(c, async (conn) => conn.query("SELECT 1"));
      return { ok: true, message: "Conexión exitosa", latencyMs: Date.now() - start };
    } catch (e) {
      return { ok: false, message: e instanceof Error ? e.message : String(e) };
    }
  },

  async listSchemas(c) {
    return withConn(c, async (conn) => {
      const [rows] = (await conn.query("SHOW DATABASES")) as [Array<unknown>, unknown];
      const SYS = new Set(["information_schema", "mysql", "performance_schema", "sys"]);
      return (rows as Array<unknown[]>).map((r) => {
        const name = String(r[0]);
        return { name, isSystem: SYS.has(name) };
      });
    });
  },

  async listRelations(c, schema) {
    return withConn(c, async (conn) => {
      const useSchema = schema ?? c.database;
      // table_rows / data_length / index_length are statistics InnoDB caches
      // and only refreshes after ANALYZE TABLE — close enough for an
      // overview, never authoritative for billing.
      const [rows] = (await conn.query(
        `SELECT table_schema, table_name, table_type,
                table_rows, data_length, index_length
         FROM information_schema.tables
         WHERE table_schema = ?
         ORDER BY table_name`,
        [useSchema],
      )) as [Array<unknown[]>, unknown];
      return rows.map((row) => {
        const [s, n, t, rowsApprox, dataLen, idxLen] = row as [
          string,
          string,
          string,
          number | string | null,
          number | string | null,
          number | string | null,
        ];
        const data = dataLen != null ? Number(dataLen) : undefined;
        const idx = idxLen != null ? Number(idxLen) : undefined;
        const total =
          data != null || idx != null ? (data ?? 0) + (idx ?? 0) : undefined;
        return {
          schema: s,
          name: n,
          kind: t === "VIEW" ? ("view" as const) : ("table" as const),
          approxRowCount: rowsApprox != null ? Number(rowsApprox) : undefined,
          totalBytes: total,
          dataBytes: data,
          indexBytes: idx,
        };
      });
    });
  },

  async getConnectionOverview(c) {
    return withConn(c, async (conn) => {
      const SYS = new Set(["information_schema", "mysql", "performance_schema", "sys"]);
      // CURRENT_USER and UPTIME are reserved/non-existent as system variables in
      // MySQL 8 — alias with backticks and read Uptime from STATUS instead of
      // `@@global.uptime` (which doesn't exist as a system variable).
      const [[meta]] = (await conn.query(
        `SELECT VERSION() AS version,
                DATABASE() AS current_database,
                CURRENT_USER() AS \`current_user\`,
                NOW() AS server_time,
                @@max_connections AS max_connections`,
      )) as [Array<unknown[]>, unknown];
      const [version, currentDb, currentUser, serverTime, maxConn] =
        (meta as unknown[]) ?? [];

      // Uptime lives in SHOW GLOBAL STATUS. Wrap so a privilege error doesn't
      // sink the whole overview.
      let uptime: number | undefined;
      try {
        const [statusRows] = (await conn.query(
          `SHOW GLOBAL STATUS LIKE 'Uptime'`,
        )) as [Array<unknown[]>, unknown];
        const row = statusRows[0] as unknown[] | undefined;
        const v = row?.[1];
        if (v != null) uptime = Number(v);
      } catch {
        /* user lacks PROCESS or similar — fine, leave uptime undefined. */
      }

      // information_schema.schemata gives the list, then aggregate sizes from
      // information_schema.tables in one round-trip.
      const [dbRows] = (await conn.query(
        `SELECT s.schema_name AS name,
                COALESCE(SUM(t.data_length + t.index_length), 0) AS size_bytes,
                COUNT(t.table_name) AS relation_count,
                s.default_character_set_name AS charset
         FROM information_schema.schemata s
         LEFT JOIN information_schema.tables t ON t.table_schema = s.schema_name
         GROUP BY s.schema_name, s.default_character_set_name
         ORDER BY s.schema_name`,
      )) as [Array<unknown[]>, unknown];

      let active: unknown;
      try {
        const [activeRows] = (await conn.query(
          `SELECT COUNT(*) AS active FROM information_schema.processlist`,
        )) as [Array<unknown[]>, unknown];
        active = (activeRows[0] as unknown[] | undefined)?.[0];
      } catch {
        /* requires PROCESS privilege — leave undefined. */
      }

      const databases = dbRows.map((row) => {
        const [name, sizeBytes, relCount, charset] = row as [
          string,
          number | string | null,
          number | string | null,
          string | null,
        ];
        const size = sizeBytes != null ? Number(sizeBytes) : undefined;
        return {
          name,
          isSystem: SYS.has(name),
          sizeBytes: size && size > 0 ? size : undefined,
          relationCount: relCount != null ? Number(relCount) : undefined,
          details: charset ?? undefined,
        };
      });
      const totalSizeBytes = databases.reduce(
        (sum, d) => sum + (d.sizeBytes ?? 0),
        0,
      );

      return {
        driver: "mysql" as const,
        serverVersion: version != null ? `MySQL ${String(version)}` : undefined,
        currentDatabase: currentDb != null ? String(currentDb) : undefined,
        currentUser: currentUser != null ? String(currentUser) : undefined,
        serverTime:
          serverTime instanceof Date
            ? serverTime.toISOString()
            : serverTime != null
            ? String(serverTime)
            : undefined,
        uptimeSeconds: uptime != null ? Number(uptime) : undefined,
        activeConnections: active != null ? Number(active) : undefined,
        maxConnections: maxConn != null ? Number(maxConn) : undefined,
        totalSizeBytes: totalSizeBytes > 0 ? totalSizeBytes : undefined,
        databases,
      };
    });
  },

  async describeTable(c, schema, name) {
    return withConn(c, async (conn) => {
      const [rows] = (await conn.query(
        `SELECT column_name, column_type, is_nullable, column_key, column_default
         FROM information_schema.columns
         WHERE table_schema = ? AND table_name = ?
         ORDER BY ordinal_position`,
        [schema, name],
      )) as [Array<unknown[]>, unknown];
      const [idxRows] = (await conn.query(
        `SELECT index_name, non_unique, column_name, seq_in_index
         FROM information_schema.statistics
         WHERE table_schema = ? AND table_name = ?
         ORDER BY index_name, seq_in_index`,
        [schema, name],
      )) as [Array<unknown[]>, unknown];
      const indexMap = new Map<string, { name: string; columns: string[]; unique: boolean; primary: boolean }>();
      for (const r of idxRows) {
        const [indexName, nonUnique, columnName] = r as [string, number | string, string, number];
        const key = indexName;
        let entry = indexMap.get(key);
        if (!entry) {
          entry = {
            name: indexName,
            columns: [],
            unique: Number(nonUnique) === 0,
            primary: indexName === "PRIMARY",
          };
          indexMap.set(key, entry);
        }
        entry.columns.push(columnName);
      }
      return {
        schema,
        name,
        kind: "table" as const,
        columns: rows.map((r) => {
          const [colName, colType, nullable, key, def] = r as [string, string, string, string, string | null];
          return {
            name: colName,
            dataType: colType,
            nullable: nullable === "YES",
            isPrimaryKey: key === "PRI",
            isUnique: key === "UNI",
            default: def,
          };
        }),
        indexes: Array.from(indexMap.values()).sort((a, b) => {
          if (a.primary !== b.primary) return a.primary ? -1 : 1;
          if (a.unique !== b.unique) return a.unique ? -1 : 1;
          return a.name.localeCompare(b.name);
        }),
        foreignKeys: [],
      };
    });
  },

  async runQuery(c, sql, options) {
    const start = Date.now();
    return withConn(c, async (conn) => {
      if (options?.schema) {
        await conn.query(`USE ${"`" + options.schema.replace(/`/g, "``") + "`"}`);
      }
      const [rows, fields] = (await conn.query({
        sql,
        values: options?.params,
      })) as [unknown, mysql.FieldPacket[] | undefined];

      // For SELECT-style queries `rows` is an array of arrays (rowsAsArray=true).
      // For DML it's an OkPacket-like object.
      if (Array.isArray(rows) && fields) {
        const columns: QueryResultColumn[] = fields.map((f) => ({
          name: f.name,
          dataType: typeName(f.type ?? -1),
        }));
        const cleaned = (rows as Array<unknown[]>).map((row) => row.map(normalizeCell));
        return {
          columns,
          rows: cleaned as QueryResult["rows"],
          rowCount: cleaned.length,
          affectedRows: null,
          durationMs: Date.now() - start,
          truncated: cleaned.length === QUERY_LIMIT,
        };
      }
      const okPacket = rows as { affectedRows?: number };
      return {
        columns: [],
        rows: [],
        rowCount: 0,
        affectedRows: okPacket.affectedRows ?? 0,
        durationMs: Date.now() - start,
        truncated: false,
      };
    });
  },

  async fetchTableData(c, schema, name, options) {
    const limit = Math.min(options?.limit ?? 100, QUERY_LIMIT);
    const offset = options?.offset ?? 0;
    const orderBy = options?.orderBy?.length
      ? `ORDER BY ${options.orderBy
          .map((o) => `${ident(o.column)} ${o.direction.toUpperCase()}`)
          .join(", ")}`
      : "";
    const where = options?.where?.trim() ? `WHERE ${options.where.trim()}` : "";
    const sql = `SELECT * FROM ${ident(schema)}.${ident(name)} ${where} ${orderBy} LIMIT ? OFFSET ?`;
    return this.runQuery(c, sql, { params: [limit, offset] });
  },

  async *iterateTable(c, opts) {
    // mysql2 supports server-side streaming via .stream(), which holds the
    // connection open and emits a row event per record. We collect into
    // fixed-size batches before yielding to keep downstream JSON encoding
    // efficient.
    const batchSize = Math.max(1, opts.batchSize ?? 1000);
    const where = opts.where?.trim() ? `WHERE ${opts.where.trim()}` : "";
    const sql = `SELECT * FROM ${ident(opts.schema)}.${ident(opts.name)} ${where}`;
    const conn = await mysql.createConnection({
      host: c.host,
      port: c.port,
      user: c.username,
      password: c.password,
      database: c.database,
      ssl: c.ssl ? {} : undefined,
      connectTimeout: 10_000,
      rowsAsArray: true,
      dateStrings: true,
      multipleStatements: false,
    });
    try {
      // The promise wrapper hides the streaming API; reach through to the
      // underlying connection where .query() returns a Query object with
      // .stream().
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const raw: any = (conn as unknown as { connection: unknown }).connection;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const stream: NodeJS.ReadableStream = raw.query(sql).stream({
        highWaterMark: batchSize,
      });
      let batchRows: unknown[][] = [];
      let columns: QueryResultColumn[] = [];
      let fields: mysql.FieldPacket[] | null = null;
      // Capture column metadata via the 'fields' event.
      (stream as NodeJS.EventEmitter).on("fields", (f: mysql.FieldPacket[]) => {
        fields = f;
        columns = f.map((field) => ({
          name: field.name,
          dataType: typeName(field.type ?? -1),
        }));
      });
      for await (const row of stream as AsyncIterable<unknown>) {
        if (columns.length === 0 && fields) {
          columns = (fields as mysql.FieldPacket[]).map((field) => ({
            name: field.name,
            dataType: typeName(field.type ?? -1),
          }));
        }
        batchRows.push((row as unknown[]).map(normalizeCell));
        if (batchRows.length >= batchSize) {
          yield { columns, rows: batchRows as QueryResult["rows"] } as TableBatch;
          batchRows = [];
        }
      }
      if (batchRows.length > 0) {
        yield { columns, rows: batchRows as QueryResult["rows"] } as TableBatch;
      }
    } finally {
      await conn.end().catch(() => undefined);
    }
  },

  generateCreateTableSql(_c, schema, name, columns) {
    return buildMysqlCreateTable(schema, name, columns);
  },
};

function buildMysqlCreateTable(
  schema: string,
  name: string,
  columns: ColumnInfo[],
): string {
  const colLines = columns.map((c) => {
    const parts = [qiCore("mysql", c.name), c.dataType];
    if (!c.nullable) parts.push("NOT NULL");
    if (c.default != null) parts.push(`DEFAULT ${c.default}`);
    return "  " + parts.join(" ");
  });
  const pk = columns.filter((c) => c.isPrimaryKey).map((c) => qiCore("mysql", c.name));
  if (pk.length > 0) {
    colLines.push(`  PRIMARY KEY (${pk.join(", ")})`);
  }
  return `CREATE TABLE ${qiCore("mysql", schema)}.${qiCore("mysql", name)} (\n${colLines.join(
    ",\n",
  )}\n);`;
}

export type { IterateTableOptions };

// MySQL field type codes -> friendly names. See mysql2 Types.
const TYPES: Record<number, string> = {
  1: "tinyint",
  2: "smallint",
  3: "int",
  4: "float",
  5: "double",
  6: "null",
  7: "timestamp",
  8: "bigint",
  9: "mediumint",
  10: "date",
  11: "time",
  12: "datetime",
  13: "year",
  15: "varchar",
  16: "bit",
  245: "json",
  246: "decimal",
  252: "blob",
  253: "varchar",
  254: "char",
};
function typeName(code: number) {
  return TYPES[code] ?? `type:${code}`;
}
